profile-widget/profilewidget2.cpp

```diff
@@ -2044,70 +2045,75 @@ void ProfileWidget2::clearPictures()
 	pictures.clear();
 }
 
-void ProfileWidget2::updatePictures(const QModelIndex &from, const QModelIndex &to)
+void ProfileWidget2::updateThumbnail(QString filename, QImage thumbnail)
 {
-	DivePictureModel *m = DivePictureModel::instance();
-	for (int picNr = from.row(); picNr <= to.row(); ++picNr) {
-		int picItemNr = picNr - m->rowDDStart;
-		if (picItemNr < 0 || (size_t)picItemNr >= pictures.size())
-			return;
-		if (!pictures[picItemNr])
-			return;
-
-		pictures[picItemNr]->setPixmap(m->index(picNr, 0).data(Qt::UserRole).value<QPixmap>());
+	auto it = std::find_if(pictures.begin(), pictures.end(), [&filename](const PictureEntry &e)
+			       { return e.filename == filename; });
+	if (it != pictures.end() && it->thumbnail) {
+		int size = Thumbnailer::defaultThumbnailSize();
+		it->thumbnail->setPixmap(QPixmap::fromImage(thumbnail.scaled(size, size, Qt::KeepAspectRatio)));
 	}
 }
 
+bool ProfileWidget2::PictureEntry::operator< (const PictureEntry &e) const
+{
+	return std::tie(offset.seconds, filename) < std::tie(e.offset.seconds, e.filename);
```

Comment:
what's the semantic of comparing two std::tie ?

Reply:
https://en.cppreference.com/w/cpp/utility/tuple/tie
> std::tie can be used to introduce lexicographical comparison to a struct or to unpack a tuple

it's the C++11 way of saying:
```
return offset.seconds < e.offet.seconds && filename < e.filename;
```

Reply:
This is C++'s idiomatic way of doing the following:

```
{
	if (offset.seconds < e.offset.seconds)
		return true;
	if (offset.seconds > e.offset.seconds)
		return false;
	return filename < e.filename;
}
```

The big advantage over hand-coding is that it is much less error-prone (talking from experience here). It also works for an arbitrary number of arguments and compiles to the same code.

Reply:
> `return offset.seconds < e.offet.seconds && filename < e.filename;`

Note that this code is not correct. If `offset.seconds < e.offset.seconds`, then the filenames will not be compared. Which just proves the point that `std:tie(...)` avoids errors. :)

Reply:
yes, makes sense.
how would the above example change if we add a third parameter to the tie?

Reply:
Then
`return std::tie(a, b,c) < std::tie(e.a, e.b, e.c)` is equivalent to
```
	if (a < e.a) return true;
	if (a > e.a) return false;
	if (b < e.b) return true;
	if (b > e.b) return false;
	return c < e.c;
```
Not sure if you can write it more concisely without `tie`.

Reply:
i guess `lexicographical comparison` is the key here. `tie` does removes the potential for written errors.